### PR TITLE
Put: Special flag handling and Cursor/Dbi alignment

### PR DESCRIPTION
**Pull request by pstutz**
The MDB_NOOVERWRITE and MDB_NODUPDATA flags now return false to communicate when the put operation was not executed (instead of slow exceptions). The semantics of Cursor.put are now fully aligned with Dbi.put. Added tests to cover both special cases.

### CODE REVIEW

**Comment by pstutz** [comment]
This PR is a first attempt at solving #39, but I did not evaluate other places where currently an exception is used to communicate information that would be more efficiently communicated with a return value.

File: src/main/java/org/lmdbjava/Cursor.java [hunk 1]
```diff
@@ -187,15 +192,17 @@ public boolean prev() {
...
    * @param key key to store
    * @param val data to store
    * @param op  options for this operation
+   * @return true if the value was put, false if MDB_NOOVERWRITE or
+   *     MDB_NODUPDATA were set and the key/value existed already.
    */
-  public void put(final T key, final T val, final PutFlags... op) {
+  public boolean put(final T key, final T val, final PutFlags... op) {
     if (SHOULD_CHECK) {
+      requireNonNull(txn);
```
Comment (benalexau) [inline src/main/java/org/lmdbjava/Cursor.java]: How could `txn` ever become null (it's checked in the constructor, and the only subsequent assignment is in `renew` and the `SHOULD_CHECK` there will pick up any attempt to make it null)?

File: src/main/java/org/lmdbjava/Cursor.java [hunk 2]
```diff
@@ -204,11 +211,20 @@ public void put(final T key, final T val, final PutFlags... op) {
...
+      if (isSet(mask, MDB_NOOVERWRITE)) {
+        kv.valOut(); // marked as in,out in LMDB C docs
+      } else if (!isSet(mask, MDB_NODUPDATA)) {
+        checkRc(rc);
+      }
+      return false;
+    } else {
+      checkRc(rc);
+      return true;
+    }
```
Comment (benalexau) [inline src/main/java/org/lmdbjava/Cursor.java]: Please remove the final `else`, as it's easier to read when the `MDB_KEYEXIST` logic is in its own distinct `if` block.

File: src/main/java/org/lmdbjava/Dbi.java [hunk 3]
```diff
@@ -304,10 +307,17 @@ public void put(final Txn<T> txn, final T key, final T val,
...
+        txn.kv().valOut(); // marked as in,out in LMDB C docs
+      } else if (!isSet(mask, MDB_NODUPDATA)) {
+        checkRc(rc);
+      }
+      return false;
+    } else {
+      checkRc(rc);
+      return true;
     }
-    checkRc(rc);
```
Comment (benalexau) [inline src/main/java/org/lmdbjava/Dbi.java]: As above, please remove the final `else`.

File: src/main/java/org/lmdbjava/MaskedFlag.java [hunk 4]
```diff
@@ -64,7 +65,9 @@ static int mask(final MaskedFlag... flags) {
    * @return true if set.
    */
   static boolean isSet(final int flags, final MaskedFlag test) {
-    requireNonNull(test);
+    if (SHOULD_CHECK) {
+      requireNonNull(test);
+    }
```
Comment (benalexau) [inline src/main/java/org/lmdbjava/MaskedFlag.java]: Please remove the `SHOULD_CHECK` condition, as it is unlikely to be faster than an inlined `requireNonNull` call. Most of the other `SHOULD_CHECK`s are guarding multiple conditions or expensive checks.

**Comment by pstutz** [comment]
Thanks for the feedback, just made the changes.

**Comment by benalexau** [comment]
Thanks for making the changes. Merged.

**Pull request merged**